File: app/src/main/java/com/example/flashcardapp/FlashcardAdapter.java
// File: FlashcardAdapter.java
package com.example.flashcardapp;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.recyclerview.widget.RecyclerView;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.concurrent.TimeUnit;

import static com.example.flashcardapp.TimeUtils.formatTimeDifference;

public class FlashcardAdapter extends RecyclerView.Adapter<FlashcardAdapter.ViewHolder> {

    private List<Flashcard> flashcards;
    private Context context;

    public FlashcardAdapter(Context context, List<Flashcard> flashcards) {
        this.context = context;
        // Sort flashcards by nextReview, with overdue ones first
        Collections.sort(flashcards, new Comparator<Flashcard>() {
            @Override
            public int compare(Flashcard f1, Flashcard f2) {
                return Long.compare(f1.getNextReview(), f2.getNextReview());
            }
        });
        this.flashcards = flashcards;
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        public TextView tvQuestion, tvAnswer, tvTimeDifference, tvInterval;

        public ViewHolder(View itemView) {
            super(itemView);
            tvQuestion = itemView.findViewById(R.id.tv_question);
            tvAnswer = itemView.findViewById(R.id.tv_answer);
            tvTimeDifference = itemView.findViewById(R.id.tv_time_difference);
            tvInterval = itemView.findViewById(R.id.tv_interval);

            // Handle row click to open the edit/delete view
            itemView.setOnClickListener(v -> {
                int position = getAdapterPosition();
                Flashcard flashcard = flashcards.get(position);
                Intent intent = new Intent(context, EditFlashcardActivity.class);
                intent.putExtra("FLASHCARD_ID", flashcard.getId());
                context.startActivity(intent);
            });
        }
    }

    @Override
    public FlashcardAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.item_flashcard, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(FlashcardAdapter.ViewHolder holder, int position) {
        Flashcard flashcard = flashcards.get(position);
        holder.tvQuestion.setText(flashcard.getQuestion());
        holder.tvAnswer.setText(flashcard.getAnswer());

        // Calculate time difference in seconds, minutes, and days
        long currentTime = System.currentTimeMillis();
        long nextReviewTime = flashcard.getNextReview();
        long timeDifferenceMillis = nextReviewTime - currentTime;

        long absTimeDifferenceMillis = Math.abs(timeDifferenceMillis);
        String timeDifferenceText = formatTimeDifference(absTimeDifferenceMillis);

        // Add a negative sign if the review is overdue
        if (timeDifferenceMillis < 0) {
            timeDifferenceText = "-" + timeDifferenceText;
        }

        long intervalValue = Math.abs(flashcard.getInterval()) * 1000L;  // Convert seconds to milliseconds

        // Set the time difference and interval values
        holder.tvTimeDifference.setText(timeDifferenceText);
        holder.tvInterval.setText(formatTimeDifference(intervalValue));
    }

    @Override
    public int getItemCount() {
        return flashcards.size();
    }
}

File: app/src/main/java/com/example/flashcardapp/main/FlashcardAdapter.kt
package com.example.flashcardapp.main

import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.flashcardapp.R
import com.example.flashcardapp.data.Flashcard
import java.util.Collections

class FlashcardAdapter(
    private val context: Context,
    private val flashcards: MutableList<Flashcard>
) : RecyclerView.Adapter<FlashcardAdapter.ViewHolder>() {

    init {
        // Sort flashcards by nextReview ascending, putting overdue first.
        Collections.sort(flashcards) { f1, f2 ->
            f1.nextReview.compareTo(f2.nextReview)
        }
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val tvQuestion: TextView = itemView.findViewById(R.id.tv_question)
        val tvAnswer: TextView = itemView.findViewById(R.id.tv_answer)
        val tvTimeDifference: TextView = itemView.findViewById(R.id.tv_time_difference)
        val tvInterval: TextView = itemView.findViewById(R.id.tv_interval)

        init {
            // Handle row click to open edit
            itemView.setOnClickListener {
                val position = adapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    val flashcard = flashcards[position]
                    val intent = Intent(context, EditFlashcardActivity::class.java)
                    intent.putExtra("FLASHCARD_ID", flashcard.id)
                    context.startActivity(intent)
                }
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(context).inflate(R.layout.item_flashcard, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val flashcard = flashcards[position]
        holder.tvQuestion.text = flashcard.question
        holder.tvAnswer.text = flashcard.answer

        val currentTime = System.currentTimeMillis()
        val nextReviewTime = flashcard.nextReview
        val timeDifferenceMillis = nextReviewTime - currentTime
        val absTimeDifferenceMillis = kotlin.math.abs(timeDifferenceMillis)
        var timeDifferenceText = TimeUtils.formatTimeDifference(absTimeDifferenceMillis)

        if (timeDifferenceMillis < 0) {
            timeDifferenceText = "-$timeDifferenceText"
        }

        val intervalValue = kotlin.math.abs(flashcard.interval) * 1000L
        holder.tvTimeDifference.text = timeDifferenceText
        holder.tvInterval.text = TimeUtils.formatTimeDifference(intervalValue)
    }

    override fun getItemCount(): Int {
        return flashcards.size
    }
}
